Assets/Scripts/Features/UniversalGroundChecker.cs

```diff
@@ -0,0 +1,118 @@
+
+        private enum GroundedState
+        {
+            GROUNDED,
+            AIRED
+        }
+
+        private bool _isOnGround;
+        private bool _isLeavingGround;
+        
+        private GroundedState _currentState;
+
+        private Vector2 _origin2D;
+        private Vector3 _origin3D;
+
+        private void FixedUpdate()
+        {
+            CalculateGroundRayCasting();
+            HandleStateTransitions();
+        }
+
+        private void CalculateGroundRayCasting()
+        {
+            _origin3D = !offSetOrTransform
+                ? transform.position + offSet3D
+                : groundCheckerTransform != null ? groundCheckerTransform.position : Vector3.zero;
+
+            _origin2D = !offSetOrTransform
+                ? (Vector2)transform.position + offSet2D
+                : groundCheckerTransform != null ? groundCheckerTransform.position : Vector2.zero;
+
+            IsGrounded = GetGround();
+        }
+
+        private bool GetGround()
+        {
+            return lineOrSphere 
+                ? is3D
+                    ? Physics.RaycastAll(_origin3D, Vector3.down, rayCastLength, groundLayer)
+                        ?.Any(collider => collider.collider.gameObject != gameObject) ?? false
+                    : Physics2D.RaycastAll(_origin2D, Vector2.down, rayCastLength, groundLayer)
+                        ?.Any(collider => collider.collider.gameObject != gameObject) ?? false
+                : is3D
+                    ? Physics.OverlapSphere(_origin3D, sphereCastRadius, groundLayer)
+                        ?.Any(collider => collider.gameObject != gameObject) ?? false
+                    : Physics2D.OverlapCircleAll(_origin2D, sphereCastRadius, groundLayer)
+                        ?.Any(collider => collider.gameObject != gameObject) ?? false;
+        }
+        
+        private void HandleStateTransitions()
+        {
+            _currentState = IsGrounded ? GroundedState.GROUNDED : GroundedState.AIRED;
+            
+            switch (_currentState)
+            {
+                case GroundedState.GROUNDED when !_isOnGround:
+                    onGroundEnter?.Invoke();
+                    _isOnGround = true;
+                    break;
+                case GroundedState.GROUNDED when !_isLeavingGround:
+                    onGroundLeave?.Invoke();
+                    _isLeavingGround = true;
+                    break;
+            }
+        }
+        
+        private void OnDrawGizmos()
+        {
+            if (!gizmos)
+                return;
+
+            Vector3 origin = is3D ? _origin3D : _origin2D;
+            Gizmos.color = gizmosColor;
+            
+            if (lineOrSphere)
+            {
+                Vector3 endPosition = origin + Vector3.down * rayCastLength;
+                Gizmos.DrawLine(origin, endPosition);
+            }
+            else
+                Gizmos.DrawWireSphere(origin, sphereCastRadius);
```

Comment:
```suggestion
            else Gizmos.DrawWireSphere(origin, sphereCastRadius);
```

Reply:
https://github.com/bas-boop/Feature-vault/wiki/Code-conventions#if-statements we don't do that here. This is done for  readability.